### PR TITLE
fix(trends): Revamp WAU/MAU query to work in all cases

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

@annikaschmid [pointed out](https://posthog.slack.com/archives/C0368RPHLQH/p1665589882835399) that WAU and MAU aggregation can only group by day, despite the UI saying otherwise.

## Changes

This adds support for all graph intervals to WAU/MAU.
Edit: Also fixes the query for sparse events (see comment below).

## How did you test this code?

~TODO: Add tests~ Added query tests.

### CODE REVIEW

**Comment by Twixes** [comment]
Initially I thought I'll just add support for intervals other than day here, but in the course of adding tests for that, I realized there is another dragon: For periods without matching events, WAU/MAU have been zero, even if the previous period was less than a week/month (respectively) before and DID have events. This resulted in unexpected sharp dropoffs and rises in graphs.
The revamp of the query was a bit bigger than I planned, but now this should be all supported and covered by tests – let me know if I missed any test case though.

**Comment by pauldambra** [comment]
I nuked my local setup this morning. And have been running make some noise

With master branch I see WAU, but with this branch none... Is that wrong?

#### master

Screenshot 2022-10-19 at 16 00 31

#### this branch

Screenshot 2022-10-19 at 15 59 35

**Comment by Twixes** [comment]
I'm unable to reproduce @pauldambra 🤔

**Comment by pauldambra** [comment]
Hmmm, I'll check again later! 🤔

**Review by mariusandra — APPROVED** [review]
works for me in master:


and in this branch:


🤔

**Comment by mariusandra** [comment]
note: I did not go deep in understanding the change in the query - let me know if I should, but on the surface it looks sensible

File: posthog/queries/trends/sql.py [hunk 1]
```diff
@@ -36,14 +36,25 @@
...
 """
 
-ACTIVE_USER_SQL = """
-SELECT counts as total, timestamp as day_start FROM (
-    SELECT d.timestamp, COUNT(DISTINCT {aggregator}) counts FROM (
-        SELECT toStartOfDay(toTimeZone(toDateTime(timestamp, 'UTC'), %(timezone)s)) as timestamp FROM events WHERE team_id = %(team_id)s {parsed_date_from_prev_range} {parsed_date_to} GROUP BY timestamp
+ACTIVE_USERS_SQL = """
+SELECT counts AS total, timestamp AS day_start FROM (
+    SELECT d.timestamp, COUNT(DISTINCT {aggregator}) AS counts FROM (
+        -- We generate a table of periods to match events against. This has to be synthesized from `numbers`
```
Comment (macobo) [inline posthog/queries/trends/sql.py:42]: Note: Comments-in-sql are kind of problematic:
- There's a max query size limit we might bump up against
- It kills copying the queries from metabase which is important

Suggestions to fix:
- Move comment above the SQL to be python code comment
- Use block comment style, not `--`

Comment (Twixes) [reply posthog/queries/trends/sql.py]: I see, thanks. I do think comments are a lot more useful next to the relevant code, so I'm wary of just piling them above the query in Python… For now I moved them to multiline `/* */` comments within the query.

File: posthog/queries/test/test_trends.py [hunk 2]
```diff
@@ -4029,79 +4032,204 @@ def _create_active_user_events(self):
...
-                distinct_id="p2",
-                timestamp="2020-01-11T12:00:00Z",
+                distinct_id="p0",
+                timestamp="2020-01-12T12:00:00Z",
                 properties={"key": "val"},
             )
 
-        def test_active_user_math(self):
-            self._create_active_user_events()
+        def test_weekly_active_users_daily(self):
```
Comment (macobo) [inline posthog/queries/test/test_trends.py:4081]: I think we should be using `@snapshot_clickhouse_queries` for this and other new tests. Also consider `@test_with_materialized_columns` if test is touching these.

Most of these predate the helper but this helps figure out the generated sql. For example I was really curious on what the shape of event_query is.

**Comment by macobo** [comment]
Didn't go deep into correctness (trusting the tests) but this change seems to make sense and is well-tested.

:shipit:

File: posthog/queries/test/test_trends.py (hunk 2 above)
Comment (Twixes) [reply posthog/queries/test/test_trends.py:4099]: Good suggestion, added `@snapshot_clickhouse_queries`

**Pull request merged**